File: BOJ/누적합/11660.py
# 11660 구간 합 구하기 5
# https://www.acmicpc.net/problem/11660
import sys
def input(): return sys.stdin.readline().rstrip()


N, M = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]
prefixsum = [[0] * (N+1) for _ in range(N+1)]

for i in range(1, N+1):
    for j in range(1, N+1):
        prefixsum[i][j] = board[i-1][j-1] + prefixsum[i-1][j] + \
            prefixsum[i][j-1] - prefixsum[i-1][j-1]

for _ in range(M):
    x1, y1, x2, y2 = map(int, input().split())
    print(prefixsum[x2][y2]-prefixsum[x1-1][y2] -
          prefixsum[x2][y1-1]+prefixsum[x1-1][y1-1])

File: BOJ/누적합/11660/11660.cpp
// 11660 구간 합 구하기 5
// https://www.acmicpc.net/problem/11660
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> v;

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    
    int N, M;
    cin >> N >> M;
    v = vector<vector<int>> (N, vector<int>(N, 0));

    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < N; ++j)
        {
            cin >> v[i][j];
        }
    }

    vector<vector<int>> prefixSum(N+1, vector<int>(N+1, 0));
    for (int i = 1; i <= N; ++i)
    {
        for (int j = 1; j <= N; ++j)
        {
            prefixSum[i][j] = v[i-1][j-1] + prefixSum[i-1][j] + prefixSum[i][j-1] - prefixSum[i-1][j-1];
        }
    }
    int sx, sy, ex, ey;
    for (int i = 0; i < M; ++i)
    {
        cin >> sx >> sy >> ex >> ey;
        cout << prefixSum[ex][ey] - prefixSum[sx-1][ey]-prefixSum[ex][sy-1]+prefixSum[sx-1][sy-1] << '\n';
    }
    return 0;
}